### DISCUSSION TITLE
Visualizations

### DISCUSSION THREAD

**Discussion by vinzenz-muser** [ZEN-universe/ZEN-garden#434] [Functionalities]
# Visualizations setup

On top of ZEN Garden, we developed a set of tools that let you explore the solutions that you created.

The implementation consists of two tools:

- A webserver (https://github.com/ZEN-universe/ZEN-temple) that uses ZEN Garden to load calculated solutions and provide the information to any consumer
- A frontend (https://github.com/ZEN-universe/ZEN-explorer) that uses said server in order to create different plots and visualizations for the solutions

# Run the visualizations

There are two possibilities to run the visualization: By running with Docker or running the programs directly.

## Run with Docker

The easiest way to run the tools locally is by using Docker. Docker is a tool that allows you to easily run applications independently of the platform and the language that they were developed with. This is done by building and running containers. Containers are virtual machines that consists of a complete operating system and an arbitrary set of tools. These containers are defined in files called `Dockerfile`.

If you have several different containers that should run together, you can write a docker-compose file which orchestrates the dependencies and building of the different containers.

In our solution, the webserver and the frontend can both be run in a container.

To use the containers, you first need to install docker:

1. Install docker (and docker desktop), following the instructions on https://docs.docker.com/get-docker/
2. Check that Docker compose has been installed correctly by running `docker compose version`
The easiest way to run both containers is by using the repository https://github.com/ZEN-universe/ZEN-land

Once you installed docker, you can follow these steps to run the visualizations tool:

1. Clone the ZEN Land repository recursively to any place that you want by running `git clone --recursive git@github.com:ZEN-universe/ZEN-land.git`. This will directly clone the necessary repositories
2. Copy the file `.env.example` in the repository to a file called `.env`and in this file adapt the path to the folder that contains your solutions
3. Build the images by running `docker compose build`
4. Run the images by running `docker compose up`

If everything worked correctly, you should see the following lines:

```
explorer-1  | 
explorer-1  | > d3explorer@0.0.1 preview
explorer-1  | > vite preview --host 0.0.0.0 --port 8050
explorer-1  | 
explorer-1  | 
explorer-1  |   ➜  Local:   http://localhost:8050/explorer
explorer-1  |   ➜  Network: http://172.22.0.3:8050/explorer
temple-1    | INFO:     Started server process [1]
temple-1    | INFO:     Waiting for application startup.
temple-1    | INFO:     Application startup complete.
temple-1    | INFO:     Uvicorn running on http://0.0.0.0:8000 (Press CTRL+C to quit)
```

This means, that the webserver ZEN Temple is running under http://localhost:8000 and the frontend under http://localhost:8050/explorer.
So in order to look at your solutions, you can go to http://localhost:8050/explorer with any browser of your choice.

> [!NOTE]
> The output of `docker` run states that the servers are running under `0.0.0.0` but we're accessing them on `localhost`. This is because we did not specify the temple host in our Explorer configuration and the default  host is `localhost`. If we went to http://0.0.0.0:8050/explorer in our browser, the visualizations would load but it could not contact the web server because it uses different domain names (`0.0.0.0` vs. `localhost`).

### Updates and different folders

If you want to update the repositories to the newest state, you can run `git submodule update --recursive --remote`. This will pull the newest changes of all the repositories. Once this is done, you will have to rebuild the images by running `docker compose build`.

If you want to change the folder of your solutions, you can simply adapt the path in the `.env` file. You do not need to rebuild the containers, simply restart the containers with `docker compose up`.

## Run directly

If you do not want to run the visualization tools with Docker, you can also run them directly. 

In order to do that you need to start the web server and the frontend independently.

### Run the Web Server (ZEN Temple)

ZEN Temple is an application written in Python that uses FastAPI.

In order to run the application locally, the following steps are necessary:

1. Clone the repository `https://github.com/ZEN-universe/ZEN-temple` and go to the cloned directory
2. (optional, Linux and Mac only) Create a virtual environment with `python -m venv venv` and activate it with `source venv/bin/activate`
3. Install requirements with `pip install -r requirements.txt`
4. Copy the file `.env.example`to a file called `.env` and adapt the path to the folder that contains the solutions
5. Run the server with `uvicorn src.main:app --reload --host 0.0.0.0`
6. (optional) Verify that everything is working by opening http://localhost:8000/docs in your browser. You can even test if the solutions are loading correctly by opening http://localhost:8000/solutions/list.

In order to update the server you can simply pull the latest changes and restart the server. If you want to change the folder of the solutions change the corresponding variable in the `.env`file and restart the service.

### Run the Frontend (ZEN Explorer)

ZEN Explorer is a Svelte app. Svelte is a Javascript-Frontend-Framework and in order to run it you need to install Node.js (https://nodejs.org/en). You can verify that Node and NPM is installed by running `npm -v`.

Once Node is installed, you can follow these steps:

1. Clone the repository `https://github.com/ZEN-universe/ZEN-explorer` and go to the cloned directory
2. Install the dependencies with `npm install`
3. Start the service by running `npm run dev`

> [!IMPORTANT]
> Keep in mind that the frontend directly communicates with the web server, so it is important to have ZEN Temple running. Otherwise no solutions will show up in the frontend, since it cannot connect to any server.

## Use Static Homepage

In order to use the statically generated frontend, follow these steps:

1. Activate a Conda-enviornment with ZEN Garden installed. If you do not have such an environment yet, follow the steps described in [How to ZEN-Garden](https://github.com/ZEN-universe/ZEN-garden/blob/main/documentation/how_to_ZEN_garden.md ).
2. Clone the repository https://github.com/ZEN-universe/ZEN-temple to any location and change to this location in your Anaconda prompt
3. Install the additional requirements by running `pip install -r requirements_conda.txt`
4. Rename the file `.env.example` to `.env`. Inside this file, change the variable `SOLUTION_FOLDER` to any path where you saved your ZEN Garden solutions. Make sure that the path uses slashes (`/`) and not backslashes (`\`).
5. Start the server by running `python explorer.py`. A webbrowser with the correct url should open automatically. Maybe you have to refresh the tab after a few seconds. If this did not work, open the url [http://localhost:8000/explorer/](http://localhost:8000/explorer/) in any browser.

**Comment by jacob-mannhardt** [comment]
@vinzenz-muser is this outdated/moved to the RTD?

**Comment by csfunke** [comment]
I am closing the discussion since it appears to be outdated.
